File: game-server/src/main/java/work/fking/pangya/game/packet/handler/room/CreateRoomPacketHandler.java
package work.fking.pangya.game.packet.handler.room;

import io.netty.buffer.ByteBuf;
import work.fking.pangya.common.Rand;
import work.fking.pangya.game.GameServer;
import work.fking.pangya.game.model.Course;
import work.fking.pangya.game.model.HoleMode;
import work.fking.pangya.game.model.RoomType;
import work.fking.pangya.game.net.ClientGamePacketHandler;
import work.fking.pangya.game.packet.outbound.RoomResponses;
import work.fking.pangya.game.player.Player;
import work.fking.pangya.networking.protocol.ProtocolUtils;

public class CreateRoomPacketHandler implements ClientGamePacketHandler {

    @Override
    public void handle(GameServer server, Player player, ByteBuf packet) {
        var unknown = packet.readByte();
        var shotTime = packet.readIntLE();
        var gameTime = packet.readIntLE();
        var maxPlayers = packet.readByte();
        var roomType = RoomType.forId(packet.readByte());
        var holeCount = packet.readByte();
        var course = Course.forId(packet.readByte());
        var holeMode = HoleMode.forId(packet.readByte());
        var unknown3 = packet.readIntLE();
        var name = ProtocolUtils.readPString(packet);
        var password = ProtocolUtils.readPString(packet);
        var artifactIffId = packet.readIntLE();

        var channel = player.channel();
        channel.write(RoomResponses.createSuccess(name, roomType, Rand.maxInclusive(127)));
        channel.write(RoomResponses.roomInfo(roomType, name, course, holeMode, holeCount, maxPlayers, shotTime, gameTime));

        channel.write(RoomResponses.roomInitialCensus(player));
        channel.write(RoomResponses.loungePkt196());
        channel.write(RoomResponses.loungePkt9e());
        channel.flush();
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/packet/handler/room/CreateRoomPacketHandler.kt
package work.fking.pangya.game.packet.handler.room

import io.netty.buffer.ByteBuf
import work.fking.pangya.game.GameServer
import work.fking.pangya.game.net.ClientPacketHandler
import work.fking.pangya.game.player.Player
import work.fking.pangya.game.room.courseById
import work.fking.pangya.game.room.holeModeById
import work.fking.pangya.game.room.roomTypeById
import work.fking.pangya.networking.protocol.readPString
import java.time.Duration
import java.time.temporal.ChronoUnit

class CreateRoomPacketHandler : ClientPacketHandler {
    override fun handle(server: GameServer, player: Player, packet: ByteBuf) {
        val unknown = packet.readByte()
        val shotTime = packet.readIntLE()
        val gameTime = packet.readIntLE()
        val maxPlayers = packet.readByte()
        val roomTypeId = packet.readByte()
        val roomType = roomTypeById(roomTypeId)
        val holeCount = packet.readByte()
        val course = courseById(packet.readByte())
        val holeModeId = packet.readByte()
        val holeMode = holeModeById(holeModeId)
        val unknown3 = packet.readIntLE()
        val name = packet.readPString()
        val password = packet.readPString()
        val artifactIffId = packet.readIntLE()

        val serverChannel = player.currentChannel ?: throw IllegalStateException("Player ${player.nickname} attempted to create a room while not in a server channel")

        val room = serverChannel.roomManager.createRoom(
            name = name,
            password = password,
            roomType = roomType,
            course = course,
            holeMode = holeMode,
            holeCount = holeCount.toInt(),
            maxPlayers = maxPlayers.toInt(),
            shotTime = Duration.of(shotTime.toLong(), ChronoUnit.MILLIS),
            gameTime = Duration.of(gameTime.toLong(), ChronoUnit.MILLIS),
            artifactIffId = artifactIffId,
            naturalWind = false
        )
        room.attemptJoin(player)
    }
}
